prime-dotnet-webapi/Services/EnrolleeAgreementService.cs

```diff
@@ -171,5 +187,24 @@ private async Task RenderHtml(params Agreement[] agreements)
                 }
             }
         }
+
+        public async Task<bool> IsOboToRuAgreementTypeChangeAsync(int enrolleeId)
+        {
+            var currentAgreementType = await GetCurrentAgreementTypeAsync(enrolleeId);
+
+            // Get only the information needed to determine expected agreement type
+            var enrollee = await _context.Enrollees
+                .AsNoTracking()
+                .Include(e => e.Certifications)
+                    .ThenInclude(c => c.License)
+                        .ThenInclude(l => l.LicenseDetails)
+                .Include(e => e.EnrolleeCareSettings)
+                .SingleOrDefaultAsync(e => e.Id == enrolleeId);
+            var agreementDto = _mapper.Map<AgreementEngineDto>(enrollee);
+
+            var expectedAgreementType = AgreementEngine.DetermineAgreementType(agreementDto);
+            return expectedAgreementType.IsValid() && currentAgreementType.IsValid() &&
```

Comment:
Why do `.IsValid()` to check if the enum produced by our system is defined when a simple null check is sufficient?

Reply:
`0` is returned by `GetCurrentAgreementTypeAsync` if no agreement associated with enrollee

Reply:
Since you are doing a select to an enum, it will default to a 0 value. Since structs can't be null. So in that case we know if it is 0 it was the default.

So we just need to have
``` expectedAgreementType != null && currentAgreementType != 0 & ...etc ```

And we can remove the IsValid method.

As well public methods should be above private and the GetCurrentAgreementTypeAsync should be a private method as it is only called within this method. you could even just move the method to within this method and only have the one method.

Reply:
Looking at the code in `GetCurrentAgreementTypeAsync`:
```C#
return await _context.Agreements
    .OrderByDescending(a => a.CreatedDate)
    .Where(a => a.EnrolleeId == enrolleeId)
    .Where(a => a.AcceptedDate != null)
    .Select(a => a.AgreementVersion.AgreementType)
   .FirstOrDefaultAsync();
```
Examining IntelleSense on the `.Select()`:
![image](https://user-images.githubusercontent.com/39168456/151638870-60d72fc4-5d6f-43af-bd8b-eb2861fd9305.png)

The enumerable returned is of type `AgreementType`, since that is what `a.AgreementVersion.AgreementType` is.
So when you do `.FirstOrDefault` on it, it returns the default value of `AgreementType`. Since enums are structs, that can't be null. The default value for enums is whatever corresponds to value `0`. But in this case, we actually _want_ null if there are no agreements that match the user. 

There are a few ways of doing this, but the simplest way to to cast the `AgreementType` to an `AgreementType?`:
`.Select(a => (AgreementType?)a.AgreementVersion.AgreementType)`. Now, the `.Select` is of type `AgreementType?` so `.FirstOrDefault` will return `null` instead of `0`.